### PR TITLE
Split AccessorDecl out from FuncDecl.  NFC.

### PR DESCRIPTION
**Pull request by rjmccall**
This has three principal advantages:

- It gives some additional type-safety when working with known accessors.

- It makes it significantly easier to test whether a declaration is an accessor and encourages the use of a common idiom.

- It saves a small amount of memory in both FuncDecl and its serialized form.

### CODE REVIEW

**Comment by rjmccall** [comment]
@swift-ci Please test.

**Comment by rjmccall** [comment]
@swift-ci Please test.

**Comment by rjmccall** [comment]
@swift-ci Please test.

**Comment by slavapestov** [comment]
Cool! You might want to run the source compat suite as well :)

**Comment by rjmccall** [comment]
@swift-ci Please test.

**Comment by rjmccall** [comment]
@swift-ci Please test.

**Comment by rjmccall** [comment]
@swift-ci Please test.

**Comment by rjmccall** [comment]
@swift-ci Please test.

**Comment by rjmccall** [comment]
@swift-ci Please test.

**Comment by rjmccall** [comment]
Looks like there's a minor update needed to LLDB.

**Comment by rjmccall** [comment]
@swift-ci Please test source compatibility.

**Comment by rjmccall** [comment]
Ugh.  I might have to revert the part of this patch that tries to set the storage declaration on the accessor during construction.  It looks like we're *extremely* lax about some of this stuff; e.g. the importer will happily allow the same accessor declaration to be the getter for two different imported properties.

**Comment by jrose-apple** [comment]
That's probably how we do inherited properties. :-(

**Comment by rjmccall** [comment]
@swift-ci Please smoke test.

**Comment by rjmccall** [comment]
apple/swift-lldb#273
@swift-ci Please smoke test.

File: include/swift/AST/Decl.h [hunk 1]
```diff
@@ -5454,6 +5421,123 @@ class FuncDecl final : public AbstractFunctionDecl,
...
+    return getAccessorKind() == AccessorKind::IsMaterializeForSet;
+  }
+  bool isAnyAddressor() const {
+    auto kind = getAccessorKind();
+    return kind == AccessorKind::IsAddressor
+        || kind == AccessorKind::IsMutableAddressor;
+  }
+
+  /// isGetterOrSetter - Determine whether this is a getter or a setter vs.
+  /// a normal function.
```
Comment (jrose-apple) [inline include/swift/AST/Decl.h]: This comment isn't correct anymore.

File: include/swift/AST/Decl.h [hunk 2]
```diff
@@ -5454,6 +5421,123 @@ class FuncDecl final : public AbstractFunctionDecl,
...
+  static bool classof(const DeclContext *DC) {
+    if (auto fn = dyn_cast<AbstractFunctionDecl>(DC))
+      return classof(fn);
+    return false;
+  }
+};
+
+inline ParameterList **FuncDecl::getParameterListBuffer() {
+  if (!isa<AccessorDecl>(this))
+    return reinterpret_cast<ParameterList**>(this+1);
```
Comment (jrose-apple) [inline include/swift/AST/Decl.h]: Can you assert that `getKind()` is actually `Func`, and that we haven't added another subclass of FuncDecl?

File: include/swift/AST/Decl.h (hunk 1 above)
Comment (rjmccall) [reply include/swift/AST/Decl.h]: Fixed.

File: include/swift/AST/Decl.h (hunk 2 above)
Comment (rjmccall) [reply include/swift/AST/Decl.h]: Sure thing.

**Comment by rjmccall** [comment]
apple/swift-lldb#273
@swift-ci Please smoke test.

**Comment by slavapestov** [comment]
Having two different getters reference the same storage seems like it would cause problems elsewhere in SILGen for example. Can we instead create a new storage for the second getter when importing?

**Comment by jrose-apple** [comment]
I think it makes sense for John to change as little behavior as possible for this commit. If it was doing the weird thing before, we should keep doing the weird thing for now and fix it in a separate PR. That way we won't have to hunt through a giant diff to find what changed.

**Comment by rjmccall** [comment]
apple/swift-lldb#273
@swift-ci Please smoke test.

**Comment by rjmccall** [comment]
apple/swift-lldb#273
@swift-ci Please smoke test.

**Comment by rjmccall** [comment]
It's actually the other way, two different storage declarations using the same getter declaration.  I'm reluctant to mess with this because there is code which explicitly tries to prevent multiple declarations with the same selector from being created in a class; that seems questionable to me (isn't it inherently import-order-sensitive?), but without understanding its purpose, I'm following Jordan's advice and trying not to upset too much.

I am still trying to make things work with the new construction order.

**Comment by rjmccall** [comment]
I did poke a hole in that code so that we can create an accessor from a method if it's already been imported as a non-accessor.

**Comment by rjmccall** [comment]
apple/swift-lldb#273
@swift-ci Please smoke test.

**Comment by rjmccall** [comment]
@DougGregor Please review the change here to test/NameBinding/scope_map.swift

**Comment by rjmccall** [comment]
apple/swift-lldb#273
@swift-ci Please smoke test.

**Comment by rjmccall** [comment]
@benlangmuir Heads up that the new parser recovery with accessors on a pattern with no obvious VarDecl seems to produce something that is arguably a regression in the syntax model.  See the test change for test/SourceKit/DocumentStructure/structure.swift.

**Comment by rjmccall** [comment]
apple/swift-lldb#273
@swift-ci Please smoke test.

**Comment by rjmccall** [comment]
apple/swift-lldb#273
@swift-ci Please smoke test.

**Comment by rjmccall** [comment]
@swift-ci Please test source compatibility.

**Comment by rjmccall** [comment]
@swift-ci Test Source Compatibility.

**Comment by rjmccall** [comment]
Source compatibility failure is:

```
  'Kitura' /Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/Kitura: error: product dependency 'KituraNet' not found
  'Kitura' /Users/buildnode/jenkins/workspace-private/swift-PR-source-compat-suite/project_cache/Kitura: error: product dependency 'KituraTemplateEngine' not found
```

**Comment by benlangmuir** [comment]
> Heads up that the new parser recovery with accessors on a pattern with no obvious VarDecl seems to produce something that is arguably a regression in the syntax model. See the test change for test/SourceKit/DocumentStructure/structure.swift.

This change is fine.  As you said, the previous recovery wasn't that great either.

**Comment by vedantk** [comment]
Hi @rjmccall, I think this broke swift-lldb/master-llvm-swift5-transition build:
https://ci.swift.org/view/master-llvm-swift5-transition/job/oss-lldb-swift-master-llvm-swift5-transition-incremental-osx/22/

Could you take a look?

**Comment by vedantk** [comment]
Ah, I see https://github.com/apple/swift-lldb/pull/273. I went ahead and cherry picked it.

**Pull request merged**